src/cursor/cur_version.c

```diff
@@ -0,0 +1,412 @@
+    cbt = (WT_CURSOR_BTREE *)table_cursor;
+    CURSOR_API_CALL(cursor, session, next, NULL);
+
+    /* The cursor should be positioned, otherwise the next call will fail. */
+    if (!F_ISSET(cursor, WT_CURSTD_KEY_SET)) {
+        WT_IGNORE_RET(__wt_msg(
+          session, "WT_ROLLBACK: rolling back version_cursor->next due to no initial position"));
+        WT_ERR(WT_ROLLBACK);
+    }
+
+    upd = version_cursor->next_upd;
+
+    if (!upd_found && !F_ISSET(version_cursor, WT_VERSION_CUR_UPDATE_EXHAUSTED)) {
+        /*
+         * If the update is an aborted update, we want to skip to the next update immediately or get
+         * the ondisk value if the update is the last one in the update chain.
+         */
+        while (upd != NULL && upd->txnid == WT_TXN_ABORTED)
+            upd = upd->next;
+
+        if (upd == NULL)
+            F_SET(version_cursor, WT_VERSION_CUR_UPDATE_EXHAUSTED);
+        else {
+            /*
+             * If the update is a tombstone, we still want to record the stop information but we
+             * also need traverse to the next update to get the full value. If the tombstone was the
+             * last update in the update list, retrieve the ondisk value.
+             */
+            version_cursor->upd_txnid = upd->txnid;
+            version_cursor->upd_durable_stop_ts = upd->durable_ts;
+            version_cursor->upd_stop_ts = upd->start_ts;
+            if (upd->type == WT_UPDATE_TOMBSTONE) {
+                upd = upd->next;
+
+                /* Make sure the next update is not an aborted update. */
+                while (upd != NULL && upd->txnid == WT_TXN_ABORTED)
+                    upd = upd->next;
+            }
+
+            if (upd == NULL)
+                F_SET(version_cursor, WT_VERSION_CUR_UPDATE_EXHAUSTED);
+            else {
+                /*
+                 * Set the version cursor's key, which contains all the record metadata for that
+                 * particular version of the update.
+                 */
+                __wt_cursor_set_key(cursor, upd->txnid, upd->start_ts, upd->durable_ts,
+                  version_cursor->upd_txnid, version_cursor->upd_stop_ts,
+                  version_cursor->upd_durable_stop_ts, upd->type, upd->prepare_state, upd->flags,
+                  WT_VERSION_UPDATE_CHAIN);
+
+                /*
+                 * Copy the update value into the version cursor as we don't know the value format.
+                 * If the update is a modify, reconstruct the value.
+                 */
+                if (upd->type != WT_UPDATE_MODIFY)
+                    __wt_upd_value_assign(cbt->upd_value, upd);
+                else
+                    WT_ERR(
+                      __wt_modify_reconstruct_from_upd_list(session, cbt, upd, cbt->upd_value));
+
+                cursor->value = cbt->upd_value->buf;
+                F_SET(cursor, WT_CURSTD_VALUE_EXT);
+                version_cursor->hs_base_upd = &(cbt->upd_value->buf);
```

Comment:
hs base upd cannot be something on the update chain

Reply:
I may have placed this in the wrong place, I meant to make the ondisk value the base update.